CRM/Contact/BAO/Contact.php

```diff
@@ -430,6 +430,48 @@ public static function &create(&$params, $fixAddress = TRUE, $invokeHooks = TRUE
     return $contact;
   }
 
+  /**
+   * Check if a contact has a name.
+   *
+   * - Individuals need a first_name or last_name
+   * - Organizations need organization_name
+   * - Households need household_name
+   *
+   * @param array $contact
+   * @return bool
+   */
+  public static function hasName(array $contact): bool {
+    $nameFields = [
+      'Individual' => ['first_name', 'last_name'],
+      'Organization' => ['organization_name'],
+      'Household' => ['household_name'],
+    ];
+    // Casting to int filters out the string 'null'
+    $cid = (int) ($contact['id'] ?? NULL);
+    $contactType = $contact['contact_type'] ?? NULL;
+    if (!$contactType && $cid) {
+      $contactType = CRM_Core_DAO::getFieldValue(__CLASS__, $cid, 'contact_type');
```

Comment:
So we can get this all down to one extra query if we retrieve the potential name fields at this point too. It's a bit messier php-wise but I'm very conscious that this is a high traffic function

Reply:
Except that `CRM_Core_DAO::getFieldValue` uses caching so if we don't use it we don't get the potential benefit of the cache.
I guess I could write a new `getFieldValues()` (plural) function which re-uses the same cache...

Reply:
Yeah - I don't think we use getFieldValue enough to actually have it cached much? I feel like we have been mostly using the api (increasingly) to retrieve stuff like that. However, I guess it's arguable enough now that I won't hold this up on it